File: iherefor-smart-development/demo/reader-app/android/app/src/main/java/com/iherefor/readerapp/data/LibraryRepository.kt
package com.iherefor.readerapp.data

import com.iherefor.readerapp.domain.Book
import com.iherefor.readerapp.domain.ReadingProgress

/** 书架仓库接口（可测 seam） */
interface LibraryRepository {
    fun books(): List<Book>
    fun add(book: Book)
    fun remove(id: String)
    fun updateProgress(bookId: String, progress: ReadingProgress)
}

/** 内存实现，按 importedAt 降序 */
class InMemoryLibraryRepository : LibraryRepository {
    private val storage = mutableListOf<Book>()

    override fun books(): List<Book> = storage.sortedByDescending { it.importedAt }

    override fun add(book: Book) { storage.add(book) }

    override fun remove(id: String) { storage.removeAll { it.id == id } }

    override fun updateProgress(bookId: String, progress: ReadingProgress) {
        val idx = storage.indexOfFirst { it.id == bookId }
        if (idx >= 0) storage[idx] = storage[idx].copy(progress = progress)
    }
}

File: iherefor-smart-development/demo/reader-app/ios/ReaderApp/Data/LibraryRepository.swift
import Foundation

/// 书架仓库协议（可测 seam）
protocol LibraryRepository {
    func books() -> [Book]
    func add(_ book: Book)
    func remove(id: String)
    func updateProgress(bookId: String, progress: ReadingProgress)
}

/// 内存实现（测试与演示用），按 importedAt 降序返回
final class InMemoryLibraryRepository: LibraryRepository {
    private var storage: [Book] = []

    func books() -> [Book] {
        storage.sorted { $0.importedAt > $1.importedAt }
    }

    func add(_ book: Book) {
        storage.append(book)
    }

    func remove(id: String) {
        storage.removeAll { $0.id == id }
    }

    func updateProgress(bookId: String, progress: ReadingProgress) {
        guard let idx = storage.firstIndex(where: { $0.id == bookId }) else { return }
        storage[idx].progress = progress
    }
}
